conn.go

```diff
@@ -373,28 +409,34 @@ func (c *conn) close() {
 
 // Err returns the connection's error state after it's been closed.
 // Calling this on an open connection will block until the connection is closed.
-func (c *conn) Err() error {
-	c.errMu.Lock()
-	defer c.errMu.Unlock()
-	return &ConnectionError{inner: c.err}
+func (c *Conn) err() error {
+	c.doneErrMu.Lock()
+	defer c.doneErrMu.Unlock()
+	return &ConnectionError{inner: c.doneErr}
 }
 
-func (c *conn) NewSession(opts *SessionOptions) (*Session, error) {
+func (c *Conn) NewSession(ctx context.Context, opts *SessionOptions) (*Session, error) {
 	c.sessionsByChannelMu.Lock()
-	defer c.sessionsByChannelMu.Unlock()
 
 	// create the next session to allocate
 	// note that channel always start at 0
 	channel, ok := c.channels.Next()
 	if !ok {
+		c.sessionsByChannelMu.Unlock()
```

Comment:
Do you mind refactoring this bit into a function so you can use `defer` for the unlock. I always get nervous I'll add an exit path and forget.

Reply:
fixed